src/rust/engine/workunit_store/src/lib.rs

```diff
@@ -371,48 +401,66 @@ impl WorkunitStore {
     inner.completed_ids.push(span_id);
   }
 
-  pub fn with_latest_workunits<F, T>(&mut self, max_verbosity: log::Level, f: F) -> T
-  where
-    F: FnOnce(&[Workunit], &[Workunit]) -> T,
-  {
+  fn handle_queued_workunits(&self) {
     let mut inner_store = self.inner.lock();
+    let receiver = self.started_workunits_rx.lock();
 
-    let should_emit = |workunit: &Workunit| -> bool { workunit.metadata.level <= max_verbosity };
+    while let Ok(started) = receiver.try_recv() {
+      Self::add_started_workunit_to_store(started, &mut inner_store);
+    }
 
-    let cur_len = inner_store.started_ids.len();
-    let latest: usize = inner_store.last_seen_started_idx;
-    let started_workunits: Vec<Workunit> = inner_store.started_ids[latest..cur_len]
-      .iter()
-      .flat_map(|id| inner_store.workunit_records.get(id))
-      .flat_map(|workunit| match workunit.state {
-        WorkunitState::Started { .. } if should_emit(&workunit) => Some(workunit.clone()),
-        WorkunitState::Started { .. } => None,
-        WorkunitState::Completed { .. } => None,
-      })
-      .map(|mut w| {
-        w.parent_id = inner_store.first_matched_parent(w.parent_id, should_emit);
-        w
-      })
-      .collect();
-    inner_store.last_seen_started_idx = cur_len;
+    let receiver = self.completed_workunits_rx.lock();
+    while let Ok((span_id, new_metadata, time)) = receiver.try_recv() {
+      Self::add_completed_workunit_to_store(span_id, new_metadata, time, &mut inner_store);
+    }
+  }
 
-    let completed_ids = &inner_store.completed_ids;
-    let cur_len = completed_ids.len();
-    let latest: usize = inner_store.last_seen_completed_idx;
-    let completed_workunits: Vec<Workunit> = inner_store.completed_ids[latest..cur_len]
-      .iter()
-      .flat_map(|id| inner_store.workunit_records.get(id))
-      .flat_map(|workunit| match workunit.state {
-        WorkunitState::Completed { .. } if should_emit(&workunit) => Some(workunit.clone()),
-        WorkunitState::Completed { .. } => None,
-        WorkunitState::Started { .. } => None,
-      })
-      .map(|mut w| {
-        w.parent_id = inner_store.first_matched_parent(w.parent_id, should_emit);
-        w
-      })
-      .collect();
-    inner_store.last_seen_completed_idx = cur_len;
+  pub fn with_latest_workunits<F, T>(&mut self, max_verbosity: log::Level, f: F) -> T
+  where
+    F: FnOnce(&[Workunit], &[Workunit]) -> T,
+  {
+    self.handle_queued_workunits();
+    let (started_workunits, completed_workunits) = {
+      let mut inner_store = self.inner.lock();
+
+      let should_emit = |workunit: &Workunit| -> bool { workunit.metadata.level <= max_verbosity };
+
+      let cur_len = inner_store.started_ids.len();
+      let latest: usize = inner_store.last_seen_started_idx;
+      let started_workunits: Vec<Workunit> = inner_store.started_ids[latest..cur_len]
```

Comment:
It feels like the queue could be used in order to avoid needing to keep the `started_id` and `completed_id` collections... instead, you could consume one entry from the queue, add to the store, call `first_matched_parent` (because the parents must already exist), and then consume the next entry off the queue?

Reply:
I don't think this can work. If we use the queue associated with the mpsc receiver as the sole source of truth for what workunits have been delivered to streaming workunit processing, then that means we can only update the workunit store when `with_latest_workunits` is called. This means that we make information that `heavy_hitters` relies upon coupled to the frequency at which we call `with_latest_workunits` (which might not be called at all, if there happens to be nothing using streaming workunits in a given pants run).

Reply:
It could work if the UI did a similar thing at the same time?

You could imagine that each "WorkunitSubscriber" did their own thing with the workunits, independently?